velox/expression/DynamicLibraryLoader.cpp

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/DynamicLibraryLoader.h"
+#include <dlfcn.h>
+#include "velox/common/base/Exceptions.h"
+
+namespace facebook::velox::exec {
+
+static constexpr const char* kSymbolName = "registry";
+
+void loadDynamicLibraryFunctions(const char* fileName) {
+  // Try to dynamically load the shared library.
+  void* handler = dlopen(fileName, RTLD_NOW);
```

Comment:
curious: what will happen if multiple `.so` is loaded ? (there will be multiple "registry" symbol right? )

Reply:
I thought the registry name is loaded from file passed so it shouldnt matter if multiple .so are loaded as long as they have different paths. Collision between functions during registrations though can happen.

Reply:
From https://pubs.opengroup.org/onlinepubs/7908799/xsh/dlopen.html, it says

> dependency ordering is used by the [dlsym()](https://pubs.opengroup.org/onlinepubs/7908799/xsh/dlsym.html) function. Load ordering is used in [dlsym()](https://pubs.opengroup.org/onlinepubs/7908799/xsh/dlsym.html) operations upon the global symbol object.

Load ordering doesn't seem to be what we want (it will always resolve to the `registry` symbol for the first one)

> Two such resolution orders are defined: load or dependency ordering. Load order establishes an ordering among symbol definitions, such that the definition first loaded (including definitions from the image file and any dependent objects loaded with it) has priority over objects added later (via dlopen()).

Dependency ordering also seems irrelevant? 

> Dependency ordering uses a breadth-first order starting with a given object, then all of its dependencies, then any dependents of those, iterating until all dependencies are satisfied.

Reply:
I can also forsee another problem where say if a .so depends on another .so which has a registry function in that case behavior can be undefined.